### PR TITLE
[Fleet] Make `.fleet-secrets` index internal

### PR DESCRIPTION
**Pull request by jillguyonnet**
### Context

Fleet team is currently implementing phase 1 of storing Fleet secrets ([meta issue](https://github.com/elastic/ingest-dev/issues/1000), [phase 1 issue](https://github.com/elastic/kibana/issues/154715)), where a a new system index was created to store the secrets (cf. [this schema](https://user-images.githubusercontent.com/1813008/218431667-079a6ca8-b046-4a88-b085-4ac522c33e9e.png)).

The access expectations are as follows:
   - Kibana system should only have write and delete access, NOT read access.
   - Fleet server should have read access only.

The `.fleet-secrets` system index was added in https://github.com/elastic/elasticsearch/pull/95625; delete privilege was added in https://github.com/elastic/elasticsearch/pull/96080.

Upon testing, Mark found out that the `kibana_system` user has all privileges on `.fleet-*` indices, thus breaking the requirements; the initial proposed fixed was to rename the index to `.secrets-fleet` in https://github.com/elastic/elasticsearch/pull/96172. Following discussion on this PR, an alternative approach was decided that will allow better clarity on the different Fleet-related system indices as well as introducing fine-grained access to `.fleet-secrets`:
1. Break down access rules for each individual index (cf. [this comment](https://github.com/elastic/elasticsearch/pull/96172#issuecomment-1551180846)). This change is implemented in https://github.com/elastic/elasticsearch/pull/97881.
2. Make the `.fleet-secrets` index internal (cf. [this comment](https://github.com/elastic/elasticsearch/pull/96172#issuecomment-1553554429)). This is what this PR does.

### What this PR does

* Make the `.fleet-secrets` system index internal.
* Implement the write, delete and read API endpoints.
* Define new cluster privileges to enforce read-only access for Kibana user and write/delete-only access for Fleet server.

### How to test this PR

The API endpoints can be tested in Kibana Dev Tools:

1. Run ES from source with `./gradlew run`.
2. Once ES is running, change the default password for the elastic user from "password" to "changeme" to match Kibana's:
   ```
   curl -k -u elastic-admin:elastic-password -H 'Content-Type: application/json' \
   http://localhost:9200/_security/user/elastic/_password -d'{"password": "changeme"}'
   curl -k -u elastic:changeme -H 'Content-Type: application/json' \
    http://localhost:9200/_security/user/kibana_system/_password -d'{"password": "changeme"}'
   ```
3. Run Kibana with `yarn start`.
4. Log into Kibana and go to `/app/dev_tools#/console`.
5. Create a secret:
   ```
   POST /_fleet/secret/
   {
     "value": "test123"
   }
   ```
6. This should succeed and return the id of the created secret, e.g.:
   ```
   {
     "id": "JoCwlokBZNKF_JG5gPBw"
   }
   ```
7. Copy the id and try to get the secret:
   ```
   GET /_fleet/secret/JoCwlokBZNKF_JG5gPBw
   ```
8. This should succeed and return the secret id and value:
   ```
   {
     "id": "JoCwlokBZNKF_JG5gPBw",
     "value": "test123"
   }
   ```
9. Try to delete the secret:
   ```
   DELETE /_fleet/secret/JoCwlokBZNKF_JG5gPBw
   ```
10. This should succeed and return the following payload:
   ```
   {
     "deleted": true
   }
   ```
11. Try to delete the same secret again. It should fail with 404 and return an error:
   ```
   {
     "error": {
       "root_cause": [
         {
           "type": "resource_not_found_exception",
           "reason": "No secret with id [JoCwlokBZNKF_JG5gPBw]"
         }
       ],
       "type": "resource_not_found_exception",
       "reason": "No secret with id [JoCwlokBZNKF_JG5gPBw]"
     },
     "status": 404
   }
   ```
12. Try to get the same secret again. It should fail with 404 and return the same error as above.

### CODE REVIEW

**Comment by jillguyonnet** [comment]
Here is a summary of my current understanding and state of this task.

#### Context

1. Fleet team is currently implementing phase 1 of storing Fleet secrets ([meta issue](https://github.com/elastic/ingest-dev/issues/1000), [phase 1 issue](https://github.com/elastic/kibana/issues/154715)), where a a new system index was created to store the secrets (cf. [this schema](https://user-images.githubusercontent.com/1813008/218431667-079a6ca8-b046-4a88-b085-4ac522c33e9e.png)).
2. The access expectations are as follows:
   - Kibana system should only have write and delete access, NOT read access.
   - Fleet server should have read access only.
 3. The `.fleet-secrets` system index was added in https://github.com/elastic/elasticsearch/pull/95625; delete privilege was added in https://github.com/elastic/elasticsearch/pull/96080.
 4. Upon testing, Mark found out that the `kibana_system` user has all privileges on `.fleet-*` indices, thus breaking the requirements; the initial proposed fixed was to rename the index to `.secrets-fleet` in https://github.com/elastic/elasticsearch/pull/96172. Following discussion on this PR, it was decided that making the index internal would suit our use case better. An internal index defines a custom REST API between the system index and the client.
 5. I have taken a first attempt at defining an interface for the internal index in this PR, leaving Mark's original branch untouched.

#### Questions

1. Local development
   - I am able to import the project locally, run `./gradlew localDistro` and `./gradlew check` on `main`.
   - I'm not sure yet how I could run ES from source to test it with a local Kibana instance, I'm guessing `./gradlew run` but I've seen errors when trying to run it.
2. High-level API for internal index
   - Based on the requirements, I assume the index needs to define read, create and delete, so I have started by adding rest handlers and actions for these (skeleton level for now, many implementation details are missing).
   - ⚠️ Will this allow us to define fine grained user access like expected? i.e. write/delete only for kibana and read only for fleet server
3. Implementation and testing: assuming the above provides a healthy basis (which I'm not sure at all at this stage 😁), I will probably need some support with some implementation details and guidelines to test locally. Ideally, it would be great if the build could be run concurrently with a local Kibana instance and a test package with secret fields and test that the secrets get written to the index.

**Comment by gwbrown** [comment]
> I'm not sure yet how I could run ES from source to test it with a local Kibana instance, I'm guessing ./gradlew run but I've seen errors when trying to run it.

You're correct that `./gradlew run` is typically the easiest way to get a running ES instance from source. Posting the error (or better yet, a build scan link) in Slack is going to be the easiest way to get help with that. It's worth noting that `./gradlew run` is restricted to single-node clusters, but that should be good enough for your purposes here.

> Based on the requirements, I assume the index needs to define read, create and delete, so I have started by adding rest handlers and actions for these (skeleton level for now, many implementation details are missing).

This is likely, but depends on exactly what you're implementing. I'm not sure what the API is intended to look like but the API skeletons you have look good, as far as what's there. I suspect the `writeTo()`/`constructor(StreamInput in)` and XContent building/parsing will probably be the biggest stumbling point; don't be afraid to ask for help there because both of those things are kind of obtuse, low-level interfaces and a massive pain to debug. However, once you get used it it's pretty straightforward, so we might be able to help with those given descriptions of the desired API requests/responses.

> Will this allow us to define fine grained user access like expected? i.e. write/delete only for kibana and read only for fleet server

The way that the Elasticsearch authorization model works is that every action (REST endpoint, approximately) has a name (which you have set to `tmpfixme`) that should look something like this: `"cluster:admin/shutdown/get"`. The `cluster` indicates that is an action that takes places at the cluster level, as opposed to an `index` action that acts on a set of indices specified in the call. After the colon separator, there's something that looks like a path. The path is matched against name patterns specified in [IndexPrivilege.java](https://github.com/elastic/elasticsearch/blob/main/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/privilege/IndexPrivilege.java) (yes, even the cluster privileges). By creating a new privilege for these actions and giving it only to the Fleet user, and not the Kibana user, any new APIs you create will be restricted to Fleet.

> Ideally, it would be great if the build could be run concurrently with a local Kibana instance and a test package with secret fields and test that the secrets get written to the index.

You can definitely run with a local Kibana instance - once you get `./gradlew run` working, just start a Kibana instance aimed at `localhost:9200`. It should work as Kibana normally would; I'm afraid I'm not sure what setting up a test package will take.

**Comment by jillguyonnet** [comment]
Thank you @gwbrown for you quick feedback! I'll come back to you on running a local cluster (will make sure that works on `main` first) and then, I'm sure, regarding the API implementation.

**Comment by williamrandolph** [comment]
It looks like the CI is failing because the descriptor doesn't have all the required fields set for an "internal" index. You may want to add a simple unit test that builds the system index descriptor for quicker iteration. Something like:

```java
import org.elasticsearch.test.ESTestCase;

import java.util.Collection;

import static org.hamcrest.Matchers.is;
import static org.hamcrest.Matchers.notNullValue;

public class FleetSystemIndexTests extends ESTestCase {
    public Collection<SystemIndexDescriptor> getSystemIndexDescriptors() {
        SystemIndexPlugin plugin = new Fleet();
        return plugin.getSystemIndexDescriptors(Settings.EMPTY);
    }

    public void testSystemIndexDescriptorFormats() {
        for (SystemIndexDescriptor descriptor : getSystemIndexDescriptors()) {
            assertTrue(descriptor.isAutomaticallyManaged());
            assertThat(descriptor.getIndexFormat().hash(), is(notNullValue()));
        }
    }
}
```

**Comment by jillguyonnet** [comment]
**WIP note:**
* Got ES running from source locally and with my local Kibana 👍 [This doc](https://github.com/elastic/kibana/blob/56e13e530d1ec995a3bbb19422921f8be4d4dba9/x-pack/plugins/monitoring/dev_docs/how_to/running_components_from_source.md) turned out to be useful.
* Got Fleet tests running locally - the commit I pushed earlier made them green but we're still missing plenty of implementation details.
* Currently working on fleshing out the API - I'll report here tomorrow on it. On that note, I checked the [requirements](https://docs.google.com/document/d/15DHwXy8bw1p1T5AutrU_vgLtlTJihrpxnE2i59hKKio/edit) and it seems to me that the following endpoints make sense:
   * `GET /_fleet/_fleet_secrets/{id}` - accessible to fleet server
   * `POST /_fleet/_fleet_secrets` - accessible to kibana user
   * `DELETE /_fleet/_fleet_secrets/{id}` - accessible to kibana user

As a reminder, the secrets have a super simple id-value structure:
```
{
	_id: 'abcd1234'
	_source: {
	   value: "abcd12345678",
      }
}
```

**Comment by jillguyonnet** [comment]
**WIP report**

I think the API implementation is now much closer to what we want, although I'm still missing some parts.

Part of what is incomplete the implementation of the `doExecute` methods of `TransportGetSecretAction`, `TransportPostSecretAction` and `TransportDeleteSecretAction`.

Also, I'm not sure yet whether I need additional mappings to pass the secret value in the create endpoint.

`./gradlew run` builds OK, so I've tried testing the create endpoint in Kibana dev tools, although I'm - expectedly, I think - blocked by authorization:

Request:
`POST /_fleet/_fleet_secrets/`

Response:
```json
{
  "error": {
    "root_cause": [
      {
        "type": "security_exception",
        "reason": "action [indices:admin/auto_create] is unauthorized for user [elastic] with effective roles [superuser] on restricted indices [.fleet-secrets], this action is granted by the index privileges [auto_configure,create_index,manage,all]"
      }
    ],
    "type": "security_exception",
    "reason": "action [indices:admin/auto_create] is unauthorized for user [elastic] with effective roles [superuser] on restricted indices [.fleet-secrets], this action is granted by the index privileges [auto_configure,create_index,manage,all]"
  },
  "status": 403
}
```

@williamrandolph @gwbrown Could you perhaps recommend a way of testing manually? I still need to work on the aforementioned `doExecute` methods, but any feedback is welcome.

cc @juliaElastic

**Comment by jillguyonnet** [comment]
Pushed some WIP IT tests as part of my efforts to make the API work 😅 

* The existing `FleetSystemIndicesIT.testCreationOfFleetSecrets` fails as can be seen from the CI.
* I added a new `FleetSecretsSystemIndexIT` test specifically for the `.fleet-secrets` API. It also mostly fails at the moment.

I left comments in these tests that detail the failures. Mainly, I think it's an access issue. This can be seen e.g. in [this build](https://gradle-enterprise.elastic.co/s/iv7rozcldu37u/tests/:x-pack:plugin:fleet:javaRestTest/org.elasticsearch.xpack.fleet.FleetSystemIndicesIT/testCreationOfFleetSecrets?top-execution=1). I don't see much more information when I run the tests locally, even with the `--info` flag. The main issue seems to be:
```
org.elasticsearch.xpack.fleet.FleetSystemIndicesIT > testCreationOfFleetSecrets FAILED
    org.elasticsearch.client.WarningFailureException: method [GET], host [http://127.0.0.1:51632], URI [.fleet-secrets/_mapping], status line [HTTP/1.1 200 OK]
    Warnings: [this request accesses system indices: [.fleet-secrets-7], but in a future major version, direct access to system indices will be prevented by default]
    {".fleet-secrets-7":{"mappings":{"dynamic":"false","_meta":{"version":"8.10.0"},"properties":{"value":{"type":"keyword","index":false}}}}}
        at __randomizedtesting.SeedInfo.seed([8D2E374A3DE39375:7B9A0EDA6EACDC4D]:0)
        at app//org.elasticsearch.client.RestClient.convertResponse(RestClient.java:343)
        at app//org.elasticsearch.client.RestClient.performRequest(RestClient.java:313)
        at app//org.elasticsearch.client.RestClient.performRequest(RestClient.java:288)
        at app//org.elasticsearch.xpack.fleet.FleetSystemIndicesIT.testCreationOfFleetSecrets(FleetSystemIndicesIT.java:191)
```

**Comment by jillguyonnet** [comment]
WIP update: I added some cluster privileges for kibana and fleet server. The corresponding tests are passing (write/delete only for kibana, read only for fleet server).

Note that I added the changes in https://github.com/elastic/elasticsearch/pull/97881 as preparation.

Aside from that, I'm still struggling with making the integration tests and manual testing work. Any input appreciated at this stage.

**Comment by jillguyonnet** [comment]
@elasticmachine merge upstream

**Comment by jillguyonnet** [comment]
WIP update: pushed some missing details and fixes. The only failures right now are in the Fleet IT:
- [FleetSystemIndicesIT.testCreationOfFleetSecrets](https://gradle-enterprise.elastic.co/s/oan5plwuks2as/tests/:x-pack:plugin:fleet:javaRestTest/org.elasticsearch.xpack.fleet.FleetSystemIndicesIT/testCreationOfFleetSecrets?top-execution=1)
- [FleetSecretsSystemIndexIT.testFleetSecretsCRUD](https://gradle-enterprise.elastic.co/s/oan5plwuks2as/tests/:x-pack:plugin:fleet:javaRestTest/org.elasticsearch.xpack.fleet.FleetSecretsSystemIndexIT/testFleetSecretsCRUD?top-execution=1) (and other tests cases as well)

All of these failures are caused by direct access to `.fleet-secrets`:
```
Warnings: [this request accesses system indices: [.fleet-secrets-7], but in a future major version, direct access to system indices will be prevented by default]
```

Interestingly, in [FleetSystemIndicesIT.testCreationOfFleetSecrets](https://github.com/elastic/elasticsearch/blob/918c861f5bc0008dd2a9d52413ed661d78f5bfe4/x-pack/plugin/fleet/src/javaRestTest/java/org/elasticsearch/xpack/fleet/FleetSystemIndicesIT.java#L184-L198) it is the two GET requests that fail, the PUT request to `.fleet-secrets` succeeds.

In the current WIP state of `FleetSecretsSystemIndexIT.testFleetSecretsCRUD`, it is a POST request that fails.

I'm sure I'm missing something silly 😂 

cc @ioanatia

File: x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/TransportPostSecretAction.java
Comment (ioanatia) [inline x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/TransportPostSecretAction.java:25]: we want to use:
```
this.client = new OriginSettingClient(client, FLEET_ORIGIN);
```
and make this change in `TransportDeleteSecretAction` too.
this should fix the failure for `org.elasticsearch.xpack.fleet.FleetSecretsSystemIndexIT.testFleetSecretsCRUD`.

**Comment by jillguyonnet** [comment]
Thanks so much @ioanatia - that did fix this problem so I can move on with this test's implementation. 🙏 

Do you know if the other IT ([the one that tests index creation](https://github.com/elastic/elasticsearch/pull/97728/files#diff-25ace8ef702d1e3eca916f5c42033a0d149bb04e0c79b63503b8477c603204d7R185)) is still relevant as the index is now internal? It's still failing.

**Comment by jillguyonnet** [comment]
WIP update: got past a good few roadblocks:
- I think `POST /_fleet/secrets` now works as expected.
- In my manual testing, `GET /_fleet/secrets/{id}` does work, but it's failing in the IT (presumably because of some access issue).
- Something's still broken with `DELETE /_fleet/secrets/{id}`, it does delete the secret but responds with `Request failed to get to the server` (correct response status).

Manual testing from Kibana's Dev Tools:

**POST**
Screenshot 2023-07-26 at 16 35 11 Screenshot 2023-07-26 at 16 35 23

**GET**
Screenshot 2023-07-26 at 16 35 41
Screenshot 2023-07-26 at 16 35 49

**DELETE**
Screenshot 2023-07-26 at 16 36 44
Screenshot 2023-07-26 at 16 36 32

@juliaElastic I was getting some help from @ioanatia who suggested that I could split the changes into multiple PRs, e.g. (in order of merging):
1. https://github.com/elastic/elasticsearch/pull/97881 (already separate)
2. Make the index internally managed + implement POST using the manage cluster privilege and not individual cluster privileges.
3. Implement GET API.
4. Implement DELETE API.
5. Implement new cluster privileges.

It feels to me that we're not super far from being there, but I could be wrong so maybe going more incremental would make sense. WDYT?

File: x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/rest/RestDeleteSecretsAction.java
Comment (jillguyonnet) [reply x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/rest/RestDeleteSecretsAction.java:47]: Yes, I thought so, couldn't figure it out as of yet but I'll give it more steam. 🙂

**Comment by juliaElastic** [comment]
> It feels to me that we're not super far from being there, but I could be wrong so maybe going more incremental would make sense. WDYT?

I agree, I think one pr is good if we can figure out delete soon, othwerise we can leave that for a follow up pr.

**Comment by jillguyonnet** [comment]
OK, so actually that fix was rather easy. By changing `DeleteSecretResponse` to implement `ToXContentObject` and using a `RestToXContentListener` I can get a payload back:
Screenshot 2023-07-27 at 10 46 01

However, I wonder if we actually need that (I haven't pushed that change). I think I was a bit put off by the error message, but the deleting and response status were correct, and I'm not sure if we need anything else than the response status to know whether something's been deleted. It makes sense to me now that the dev tools were giving an error since I wasn't returning xcontent on delete.

@juliaElastic @ioanatia Thoughts?

**Comment by juliaElastic** [comment]
I think returning a `{ "deleted": true }` is nice to have, as it follows the convention of ES.

**Comment by jillguyonnet** [comment]
Pushed 👍  And I amended the testing steps in the PR steps to show the responses.

**Review by juliaElastic — APPROVED** [review]
LGTM, thanks for working through this! 🚀 
Good to merge when ES team approves.

**Comment by ioanatia** [comment]
I can take a closer look at the code changes today - in the meantime, the PR is missing yml REST tests which we usually need to add when adding new APIs.

https://github.com/elastic/elasticsearch/blob/main/TESTING.asciidoc#testing-the-rest-layer

more concrete examples:
- how to set the rest tests https://github.com/elastic/elasticsearch/pull/94011
- it's easier to create a yml test file for each API, for example https://github.com/elastic/elasticsearch/tree/main/x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch

**Comment by jillguyonnet** [comment]
I pushed some WIP yaml REST tests as I'm not sure yet how to make them work but I wanted to show the general idea. My main issue is getting the generated id back. TBC 🙂

**Comment by jillguyonnet** [comment]
Yaml REST tests are passing now 👌 I'm not sure the one about the POST endpoint is of much value as there is redundancy between them.

**Review by kc13greiner — APPROVED** [review]
LGTM!

**Comment by juliaElastic** [comment]
Hi @williamrandolph, we are planning to merge this pr soon, let me know if you want to add anything to the review.

File: x-pack/docs/en/rest-api/security/get-service-accounts.asciidoc
Comment (ioanatia) [inline x-pack/docs/en/rest-api/security/get-service-accounts.asciidoc:69]: as a follow up, do we want to provide additional documentation?

for example:
- document the privileges in https://www.elastic.co/guide/en/elasticsearch/reference/master/security-privileges.html
- add docs pages for the new APIs in  https://www.elastic.co/guide/en/elasticsearch/reference/master/fleet-apis.html

File: x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/TransportDeleteSecretAction.java
Comment (ioanatia) [inline x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/TransportDeleteSecretAction.java:38]: I am not sure why this is needed - we don't really need to do a refresh here because we are not searching over the fleet secrets index, we do a GET document request for the GET secret API and AFAIK GET document does not need a refresh to return the latest value.
(same comment for POST)

File: x-pack/plugin/fleet/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/fleet/30_secrets_post.yml
Comment (ioanatia) [inline x-pack/plugin/fleet/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/fleet/30_secrets_post.yml:1]: for the yml tests, we might want to add additional tests with a role that does not have the `read_fleet_secret`/`write_fleet_secret` privileges.

**Review by ioanatia — COMMENTED** [review]
looks good overall - great progress! 🎉 

as a follow up we might want to have additional validation on the POST API to make `value` required.

examples while testing:

```
POST _fleet/secrets
{
  "something": "else"
}
it's also possible to do:

POST _fleet/secrets
{
}
```

then when testing with the GET API, we return a 500:
```
GET _fleet/secrets/h2Ukq4kBr3pB1VkH0csR

{
  "error": {
    "root_cause": [
      {
        "type": "null_pointer_exception",
        "reason": """Cannot invoke "Object.toString()" because the return value of "java.util.Map.get(Object)" is null"""
      }
    ],
    "type": "null_pointer_exception",
    "reason": """Cannot invoke "Object.toString()" because the return value of "java.util.Map.get(Object)" is null"""
  },
  "status": 500
}
```

File: rest-api-spec/src/main/resources/rest-api-spec/api/fleet.get_secret.json
Comment (tvernum) [inline rest-api-spec/src/main/resources/rest-api-spec/api/fleet.get_secret.json:15]: The [ES API conventions](https://docs.elastic.dev/elasticsearch-team/api-guidelines/naming#choosing-the-request-path) recommend that this should be singular.
```suggestion
          "path":"/_fleet/secret/{id}",
```

File: x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/rest/RestDeleteSecretsAction.java
Comment (tvernum) [inline x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/rest/RestDeleteSecretsAction.java:20]: If these APIs are supposed to be available in Serverless (which I assume they are) then they need to be annotated
```suggestion
@ServerlessScope(INTERNAL)
public class RestDeleteSecretsAction extends BaseRestHandler {
```

I assume they're `INTERNAL` - that is, only fleet-server & kibana are supposed to be able to access them, no external clients.

**Review by tvernum — COMMENTED** [review]
I left a couple of minor comments

File: x-pack/docs/en/rest-api/security/get-service-accounts.asciidoc
Comment (juliaElastic) [reply x-pack/docs/en/rest-api/security/get-service-accounts.asciidoc:69]: Do you know where to raise a request to update ES docs?

Comment (ioanatia) [reply x-pack/docs/en/rest-api/security/get-service-accounts.asciidoc:69]: @tvernum mentioned that these APIs are internal and will only be used by Kibana and the fleet server.
in that case there might be no urgency in documenting them, so I think it's fine to defer this @juliaElastic .

File: x-pack/plugin/fleet/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/fleet/30_secrets_post.yml
Comment (juliaElastic) [reply x-pack/plugin/fleet/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/fleet/30_secrets_post.yml:1]: is there an example how to set up a role that doesn't have those privileges?

File: x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/PostSecretRequest.java
Comment (juliaElastic) [inline x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/PostSecretRequest.java:53]: @ioanatia added validation here, it might be better to actually parse the source as json first, WDYT?

File: x-pack/plugin/fleet/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/fleet/30_secrets_post.yml
Comment (ioanatia) [reply x-pack/plugin/fleet/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/fleet/30_secrets_post.yml:1]: yes, so there's 2 approaches we can take here:

A) https://github.com/elastic/elasticsearch/pull/97853 - the Elasticsearch users are created via API in yml test files themselves - we just need to take care to delete them in the teardown section

B) or we can define a `roles.yml` like in https://github.com/elastic/elasticsearch/pull/94011 - this means there's no need to do always setup the users and roles in the setup section of each file. the roles and the defined users can be reused in yml files.
For this we need to add a roles.yml file - e.g. this is how you would define an unprivileged role:

https://github.com/elastic/elasticsearch/blob/faa004df2f1f42688656a14f972d8f1f7deb4b0e/x-pack/plugin/ent-search/qa/rest/roles.yml#L28-L35

then we need to change build.gradle:
https://github.com/elastic/elasticsearch/blob/faa004df2f1f42688656a14f972d8f1f7deb4b0e/x-pack/plugin/ent-search/qa/rest/build.gradle#L14-L22

and then this is how we would test with the user that has the `unprivileged` role:

https://github.com/elastic/elasticsearch/blob/faa004df2f1f42688656a14f972d8f1f7deb4b0e/x-pack/plugin/ent-search/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/entsearch/40_search_application_delete.yml#L24-L37

I think either one is fine, although the roles.yml approach is more prevalent in xpack plugins.

File: x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/PostSecretRequest.java
Comment (ioanatia) [reply x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/PostSecretRequest.java]: yes, we want to parse the JSON.



An example:

We need a PARSER in the `PostSecretRequest` and a static method `parseRequest` method, like here:

https://github.com/elastic/elasticsearch/blob/295bdbf9e6e1f5232f32821c2d5c3c76a834ec4c/modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/action/PutDataStreamLifecycleAction.java#L53-L64

then in the rest handler, this is how the request object is created:

https://github.com/elastic/elasticsearch/blob/295bdbf9e6e1f5232f32821c2d5c3c76a834ec4c/modules/data-streams/src/main/java/org/elasticsearch/datastreams/lifecycle/rest/RestPutDataStreamLifecycleAction.java#L42

in our case, we would need to call that static parse method here:

https://github.com/elastic/elasticsearch/blob/827e9c4fe4afed945925235236b906390dbb6dfa/x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/rest/RestPostSecretsAction.java#L40

Comment (juliaElastic) [reply x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/PostSecretRequest.java]: @ioanatia I tried this but I can't figure it out. See this commit: https://github.com/elastic/elasticsearch/pull/97728/commits/acf4835725488b6d1a9cba0b79cb480069070cb8
When I add the parser, and set `PostSecretRequest.source` as the secret value, it fails at [TransportPostSecretAction.doExecute](https://github.com/elastic/elasticsearch/blob/f299539adbbe5d2a2164f6dd53d01ffce64e9e14/x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/TransportPostSecretAction.java#L34) because source is expected to be a JSON. So I would need to transform the value back to JSON, which I tried with `XContentBuilder` but I didn't find how to get the string representation from it.

```
    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
        builder.startObject();
        builder.field(VALUE_FIELD.getPreferredName(), this.value);
        builder.endObject();
        return builder;
    }
```

**Comment by juliaElastic** [comment]
Jenkins, test this

File: x-pack/plugin/fleet/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/fleet/30_secrets_post.yml
Comment (juliaElastic) [reply x-pack/plugin/fleet/qa/rest/src/yamlRestTest/resources/rest-api-spec/test/fleet/30_secrets_post.yml:17]: I'm also trying to add this, and running into problems, see commit https://github.com/elastic/elasticsearch/pull/97728/commits/1137a5d06936fcd9db4bc075927b23f903c908ab

>    headers: { Authorization: "Basic ZW50c2VhcmNoLXVucHJpdmlsZWdlZDplbnRzZWFyY2gtdW5wcml2aWxlZ2VkLXVzZXI=" }  # unprivileged 

Where is this authorization value coming from? I found the [transform_and_set](https://github.com/elastic/elasticsearch/blob/main/rest-api-spec/src/yamlRestTest/resources/rest-api-spec/test/README.asciidoc#transform_and_set-1) feature, but I'm getting an error when trying to use it.

```
java.lang.RuntimeException: Failure at [fleet/40_secrets_get:35]: Cannot invoke "org.elasticsearch.test.rest.yaml.ClientYamlTestResponse.evaluate(String, org.elasticsearch.test.rest.Stash)" because "this.response" is null
```

File: x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/PostSecretRequest.java
Comment (ioanatia) [reply x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/PostSecretRequest.java]: This is a common pattern that's being used to parse the content and to have required parameters.
We need validation here to require `value` and to make sure `value` is a string.
It's okay with me if we want to follow up on this in another PR. Same with the privilege yml tests.

You are right with this change `PostSecretRequest.source` is actually the `value` field, but then we use it here as a JSON:

https://github.com/elastic/elasticsearch/blob/731cd6ff915b20ec9f78769c3d7e0a5204209aa8/x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/TransportPostSecretAction.java#L34

so this would need a bit more refactoring in order to work, and it's perfectly fine to me to do it separately.

Comment (juliaElastic) [reply x-pack/plugin/fleet/src/main/java/org/elasticsearch/xpack/fleet/action/PostSecretRequest.java]: okay, reverted to the last passing commit and will add the changes to a new pr

**Review by ioanatia — APPROVED** [review]

**Review by williamrandolph — COMMENTED** [review]
Sorry for the late review, but this looks very good to me. Great work figuring out all the pieces here!

**Pull request merged**
